core/src/impl/Kokkos_Combined_Reducer.hpp

```diff
@@ -0,0 +1,670 @@
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_COMBINED_REDUCER_HPP
+#define KOKKOS_COMBINED_REDUCER_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Core_fwd.hpp>
+
+#include <Kokkos_Parallel_Reduce.hpp>
+#include <Kokkos_ExecPolicy.hpp>
+#include <Kokkos_AnonymousSpace.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+// TODO move this to a more general backporting facilities file
+
+// acts like void for comma fold emulation
+struct _fold_comma_emulation_return {};
+
+template <class... Ts>
+KOKKOS_INLINE_FUNCTION void emulate_fold_comma_operator(Ts&&...) noexcept {}
+
+//==============================================================================
+// <editor-fold desc="CombinedReducer reducer and value storage helpers"> {{{1
+
+// Note: the index is only to avoid repeating the same base class multiple times
+template <size_t /*Idx*/, class ValueType>
+struct CombinedReducerValueItemImpl {
+ public:
+  using value_type = ValueType;
+
+ private:
+  value_type m_value;
+
+ public:
+  KOKKOS_DEFAULTED_FUNCTION constexpr CombinedReducerValueItemImpl() = default;
+  KOKKOS_DEFAULTED_FUNCTION constexpr CombinedReducerValueItemImpl(
+      CombinedReducerValueItemImpl const&) = default;
+  KOKKOS_DEFAULTED_FUNCTION constexpr CombinedReducerValueItemImpl(
+      CombinedReducerValueItemImpl&&) = default;
+  KOKKOS_DEFAULTED_FUNCTION KOKKOS_CONSTEXPR_14 CombinedReducerValueItemImpl&
+  operator=(CombinedReducerValueItemImpl const&) = default;
+  KOKKOS_DEFAULTED_FUNCTION KOKKOS_CONSTEXPR_14 CombinedReducerValueItemImpl&
+  operator=(CombinedReducerValueItemImpl&&) = default;
+  ~CombinedReducerValueItemImpl()           = default;
```

Comment:
hipcc does not like this line `error: Destructor's restriction specifiers must cover the union of restrictions on all constructors`

Reply:
See https://github.com/ROCm-Developer-Tools/HIP/issues/660.

Reply:
Looks like just adding `KOKKOS_DEFAULTED_FUNCTION` (which should be there anyway) fixed the problem